src/GitHub.VisualStudio/Services/UsageTracker.cs

```diff
@@ -1,319 +1,105 @@
-            SaveUsage(usage);
-        }
-
-        public async Task IncrementPRDetailsViewChanges()
-        {
-            var usage = await LoadUsage();
-            ++usage.Model.NumberOfPRDetailsViewChanges;
-            SaveUsage(usage);
-        }
-
-        public async Task IncrementPRDetailsViewFile()
-        {
-            var usage = await LoadUsage();
-            ++usage.Model.NumberOfPRDetailsViewFile;
-            SaveUsage(usage);
-        }
-
-        public async Task IncrementPRDetailsCompareWithSolution()
-        {
-            var usage = await LoadUsage();
-            ++usage.Model.NumberOfPRDetailsCompareWithSolution;
-            SaveUsage(usage);
-        }
-
-        public async Task IncrementPRDetailsOpenFileInSolution()
-        {
-            var usage = await LoadUsage();
-            ++usage.Model.NumberOfPRDetailsOpenFileInSolution;
-            SaveUsage(usage);
-        }
-
-        public async Task IncrementPRReviewDiffViewInlineCommentOpen()
+        public async Task IncrementCounter(Expression<Func<UsageModel, int>> counter)
         {
             var usage = await LoadUsage();
-            ++usage.Model.NumberOfPRReviewDiffViewInlineCommentOpen;
-            SaveUsage(usage);
+            var property = (MemberExpression)counter.Body;
+            var propertyInfo = (PropertyInfo)property.Member;
+            var value = (int)propertyInfo.GetValue(usage.Model);
+            propertyInfo.SetValue(usage.Model, value + 1);
+            await service.WriteLocalData(usage);
         }
 
-        public async Task IncrementPRReviewDiffViewInlineCommentPost()
+        IDisposable StartTimer()
         {
-            var usage = await LoadUsage();
-            ++usage.Model.NumberOfPRReviewDiffViewInlineCommentPost;
-            SaveUsage(usage);
+            return service.StartTimer(TimerTick, TimeSpan.FromMinutes(3), TimeSpan.FromHours(8));
         }
 
         async Task Initialize()
         {
             // The services needed by the usage tracker are loaded when they are first needed to
             // improve the startup time of the extension.
-            if (userSettings == null)
+            if (!initialized)
             {
                 await ThreadingHelper.SwitchToMainThreadAsync();
 
                 client = gitHubServiceProvider.TryGetService<IMetricsService>();
                 connectionManager = gitHubServiceProvider.GetService<IConnectionManager>();
                 userSettings = gitHubServiceProvider.GetService<IPackageSettings>();
                 vsservices = gitHubServiceProvider.GetService<IVSServices>();
-
-                var program = gitHubServiceProvider.GetService<IProgram>();
-                storePath = System.IO.Path.Combine(
-                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
-                    program.ApplicationName,
-                    StoreFileName);
+                initialized = true;
```

Comment:
Is there a possibility of a race condition here where `await Initialize` gets called at the same time?

Reply:
There is a small possibility, if a counter is initialized on a background thread and while switching to the main thread, another counter is initialized. However all that would happen in that scenario is that services would get loaded twice, which I didn't think was a big enough problem to warrant putting a `SemaphoreSlim` or similar around.